File: exemplesPAP/test/tp00/TestMyMath.java
package tp00;


import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import org.junit.jupiter.api.Test;

class TestMyMath {
	@Test
	void testFactorielle() {
		assertEquals(120, MyMath.factorielle(5));
		assertEquals(1, MyMath.factorielle(0));
		assertEquals(1, MyMath.factorielle(1));
		assertEquals(2, MyMath.factorielle(2));
		assertThrows(AssertionError.class, () -> MyMath.factorielle(-1));
		assertThrows(AssertionError.class, () -> MyMath.factorielle(21));
	}

	@Test
	void testPower1() {
		assertEquals(1.0, MyMath.power(0.0, 0));
		assertEquals(1, MyMath.power(-1, 0));
		assertEquals(0, MyMath.power(0.0, 4));
		assertEquals(4, MyMath.power(4.0, 1));
		assertEquals(-8.0, MyMath.power(-2.0, 3));
		assertEquals(4.0, MyMath.power(-2.0, 2));
		assertEquals(125, MyMath.power(5.0, 3));
		assertEquals(2.5961484292674138E33, MyMath.power(2, 111));
		assertThrows(AssertionError.class, () -> MyMath.power(0, -1));
	}

	@Test
	void testPgcd() {
		assertEquals(3, MyMath.pgcd(15, 9));
		assertEquals(3, MyMath.pgcd(9, 15));
		assertEquals(10, MyMath.pgcd(10, 10));
		assertEquals(1, MyMath.pgcd(17, 31));
		assertThrows(AssertionError.class, () -> MyMath.pgcd(0, -1));
		assertThrows(AssertionError.class, () -> MyMath.pgcd(-4, 0));
	}

	@Test
	void testNbUn() {
		assertEquals(0, MyMath.nbUn(0));
		assertEquals(1, MyMath.nbUn(1));
		assertEquals(1, MyMath.nbUn(8));
		assertEquals(8, MyMath.nbUn(255));
		assertEquals(64, MyMath.nbUn(-1));
	}
	@Test
	void testPower2() {
		assertEquals(1.0, MyMath.power2(0.0, 0));
		assertEquals(1, MyMath.power2(-1, 0));
		assertEquals(0, MyMath.power2(0.0, 4));
		assertEquals(4, MyMath.power2(4.0, 1));
		assertEquals(-8.0, MyMath.power2(-2.0, 3));
		assertEquals(4.0, MyMath.power2(-2.0, 2));
		assertEquals(125, MyMath.power2(5.0, 3));
		assertEquals(2.5961484292674138E33, MyMath.power(2, 111));
		assertThrows(AssertionError.class, () -> MyMath.power(0, -1));
	}
	
	@Test
	void testUtf8() {
		assertEquals(1, MyMath.nbBytesUtf8('A'));
		assertEquals(2, MyMath.nbBytesUtf8('é'));
		assertEquals(3, MyMath.nbBytesUtf8('€'));
	}
	

}

File: exemplesPAP/python/test_mymath.py
import unittest
import mymath

class TestMyMath(unittest.TestCase):
    
    def test_fact(self):
        self.assertEqual(mymath.factorielle(5),120)
        self.assertEqual(mymath.factorielle(0),1)
        self.assertEqual(mymath.factorielle(1),1)
        self.assertEqual(mymath.factorielle(2),2)
        self.assertRaises(AssertionError,mymath.factorielle,-1)
        self.assertRaises(AssertionError,mymath.factorielle,2.3)
        self.assertRaises(AssertionError,mymath.factorielle,'B')
        
    def test_power(self):
        self.assertEqual(mymath.power( 2, y = 3),8)
        self.assertEqual(mymath.power(5.0),25)
        self.assertEqual(mymath.power(2.0,3),8)
        self.assertEqual(mymath.power(2.0,0),1)
        self.assertEqual(mymath.power(0.0,0),1)
        self.assertEqual(mymath.power(-1.0,0),1)
        self.assertEqual(mymath.power(8.0,1),8)
        self.assertEqual(mymath.power(-2.0,3),-8)
        self.assertEqual(mymath.power(-2,3),-8)
        self.assertEqual(mymath.power(2.0,111),2.0**111)
        #self.assertRaises(AssertionError,mymath.power,2,3)
        self.assertRaises(AssertionError,mymath.power,2.0,3.1)
        self.assertRaises(AssertionError,mymath.power,2.0,'A')
        self.assertRaises(AssertionError,mymath.power,2.0,-5)    
    
    def test_power2(self):
        self.assertEqual(mymath.power2(2,1),2)
        self.assertEqual(mymath.power2(2.0,3),8)
        self.assertEqual(mymath.power2(2.0,0),1)
        self.assertEqual(mymath.power2(0.0,0),1)
        self.assertEqual(mymath.power2(-1.0,0),1)
        self.assertEqual(mymath.power2(8.0,1),8)
        self.assertEqual(mymath.power2(-2.0,3),-8)
        self.assertEqual(mymath.power2(2.0,111),2.0**111)
        self.assertEqual(mymath.power2(2,-1),0.5)
        self.assertEqual(mymath.power2(-2,-3),-0.125)
        self.assertEqual(mymath.power2(-1,-1),-1)
        self.assertEqual(mymath.power2(1,-5),1)
        self.assertEqual(mymath.power2(-2,-2),(-2)**-2)
        self.assertRaises(AssertionError,mymath.power2,2.0,3.1)
        self.assertRaises(AssertionError,mymath.power2,2.0,'A')
        self.assertRaises(ZeroDivisionError,mymath.power2,0,-1)
        
    def test_pgcd(self):
        self.assertEqual(mymath.pgcd(5,5),5)
        self.assertEqual(mymath.pgcd(15,9),3)
        self.assertEqual(mymath.pgcd(9,15),3)
        self.assertEqual(mymath.pgcd(23,19),1)
        self.assertRaises(AssertionError,mymath.pgcd,7,0)
        self.assertRaises(AssertionError,mymath.pgcd,-7,7)
        self.assertRaises(AssertionError,mymath.pgcd,0,0)
        self.assertRaises(AssertionError,mymath.pgcd,5.0,6.0)
    
    def test_nbUn(self):
        self.assertEqual(mymath.nbUn(0),0)
        self.assertEqual(mymath.nbUn(1),1)
        self.assertEqual(mymath.nbUn(8),1)
        self.assertEqual(mymath.nbUn(255),8)
        self.assertRaises(AssertionError,mymath.nbUn,-1)
        self.assertRaises(AssertionError,mymath.nbUn,'ABC')
        
    def test_Utf8(self):
        self.assertEqual(mymath.nbBytesUtf8('A'),1)
        self.assertEqual(mymath.nbBytesUtf8('é'),2)
        self.assertEqual(mymath.nbBytesUtf8('€'),3)
        self.assertRaises(AssertionError,mymath.nbBytesUtf8,'ABC')
        self.assertRaises(AssertionError,mymath.nbBytesUtf8,2)
        
    def test_Utf8_V2(self):
        self.assertEqual(mymath.nbBytesUtf8_V2('A'),1)
        self.assertEqual(mymath.nbBytesUtf8_V2('é'),2)
        self.assertEqual(mymath.nbBytesUtf8_V2('€'),3)
        self.assertRaises(AssertionError,mymath.nbBytesUtf8_V2,'ABC')
        self.assertRaises(AssertionError,mymath.nbBytesUtf8_V2,2)

if __name__ == "__main__":
    unittest.main()